File: addons/guide-cpp/src/modifiers/guide_modifier_negate.cpp
#include "guide_modifier_negate.h"

using namespace godot;

void GUIDEModifierNegate::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_x"), &GUIDEModifierNegate::get_x);
    ClassDB::bind_method(D_METHOD("set_x", "value"), &GUIDEModifierNegate::set_x);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "x"), "set_x", "get_x");

    ClassDB::bind_method(D_METHOD("get_y"), &GUIDEModifierNegate::get_y);
    ClassDB::bind_method(D_METHOD("set_y", "value"), &GUIDEModifierNegate::set_y);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "y"), "set_y", "get_y");

    ClassDB::bind_method(D_METHOD("get_z"), &GUIDEModifierNegate::get_z);
    ClassDB::bind_method(D_METHOD("set_z", "value"), &GUIDEModifierNegate::set_z);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "z"), "set_z", "get_z");
}

GUIDEModifierNegate::GUIDEModifierNegate() {
    _update_caches();
}

GUIDEModifierNegate::~GUIDEModifierNegate() {
}

void GUIDEModifierNegate::_update_caches() {
    _multiplier.x = x ? -1.0 : 1.0;
    _multiplier.y = y ? -1.0 : 1.0;
    _multiplier.z = z ? -1.0 : 1.0;
}

Vector3 GUIDEModifierNegate::_modify_input(Vector3 input, double delta, int value_type) const {
    if (!input.is_finite()) return Vector3(Math_INF, Math_INF, Math_INF);
    return input * _multiplier;
}

bool GUIDEModifierNegate::is_same_as(const Ref<GUIDEModifier> &other) const {
    Ref<GUIDEModifierNegate> o = other;
    if (o.is_null()) return false;
    return o->get_x() == x && 
           o->get_y() == y &&
           o->get_z() == z;
}

String GUIDEModifierNegate::_editor_name() const {
    return "Negate";
}

String GUIDEModifierNegate::_editor_description() const {
    return "Inverts input per axis.";
}

File: addons/GuideCS/Modifiers/GUIDEModifierNegate.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEModifierNegate : GUIDEModifier {
	private bool _x = true;
	[Export]
	public bool X {
		get => _x;
		set {
			if (_x == value) return;
			_x = value;
			_UpdateCaches();
			EmitChanged();
		}
	}

	private bool _y = true;
	[Export]
	public bool Y {
		get => _y;
		set {
			if (_y == value) return;
			_y = value;
			_UpdateCaches();
			EmitChanged();
		}
	}

	private bool _z = true;
	[Export]
	public bool Z {
		get => _z;
		set {
			if (_z == value) return;
			_z = value;
			_UpdateCaches();
			EmitChanged();
		}
	}

	private Vector3 _multiplier = new Vector3(-1, -1, -1);

	public override bool IsSameAs(GUIDEModifier other) {
		return other is GUIDEModifierNegate otherNeg &&
			   X == otherNeg.X &&
			   Y == otherNeg.Y &&
			   Z == otherNeg.Z;
	}

	private void _UpdateCaches() {
		_multiplier = new Vector3(X ? -1 : 1, Y ? -1 : 1, Z ? -1 : 1);
	}

	public override Vector3 _ModifyInput(Vector3 input, float delta, GUIDEAction.GUIDEActionValueType valueType) {
		if (!input.IsFinite()) return Vector3.Inf;
		return input * _multiplier;
	}

	public override string _EditorName() => "Negate";

	public override string _EditorDescription() => "Inverts input per axis.";
}
